docs/contracts.rst

```diff
@@ -129,7 +129,7 @@ Each Contract Factory exposes the following methods.
 
     Returns the transaction hash for the deploy transaction.
 
-.. py:classmethod:: Contract.eventFilter(event_name, filter_params=None)
+.. py:classmethod:: Contract.events.<event name>.createFilter(filter_params=None)
```

Comment:
Just realizing we're using a legacy API that was never really re-evaluated for whether we could improve it.


What if instead of `filter_params` as a single argument you could pass in all of the individual values allowed in filter params, probably as keyword arguments.

```python
c.events.Transfer.createFilter(topics=..., args=..., _from=..., fromBlock=..., toBlock=...)
```

Note that `_from` kind of awkward due to the collision with the `from` keyword in python so maybe we could even change that one to `address=...`?

Reply:
What if we kept the filter param dict as an option to give raw arguments, and had the new kwargs do some argument processing.  For example, if providing `topics` through the keyword arg, you wouldnt need to include the event signature as the first arg and could enter the other topics as strings or integers values, rather than hex string representations.

Also, the AND or OR syntax for topics is a little odd:

from the json-rpc doc:

>A note on specifying topic filters:

>Topics are order-dependent. A transaction with a log with topics [A, B] will be matched by the following topic filters:

>   [] "anything"
>   [A] "A in first position (and anything after)"
>   [null, B] "anything in first position AND B in second position (and anything after)"
>   [A, B] "A in first position AND B in second position (and anything after)"
>   [[A, B], [A, B]] "(A OR B) in first position AND (A OR B) in second position (and anything after)"

Can you think of a clearer syntax? Maybe a single string of topics with "," for AND (e.g, followed by...) "|" for OR and "*" for ANY?  The above examples would look like:

"*"
"A"
"*, B"
"A, B"
"A|B, A|B"

Im not sure that this is an improvement. It would be cool if the "and anything after" matching could be controlled with the syntax, but I dont think there is a way to have "and nothing after".

Reply:
Maybe rather than a DLS (which is what the string based version you proposed) we could do this with a custom helper type.

```
>>> t(A) & t(B)) , t(A) | t(B)
[[A, B], [A, B]]
```

This fancy new `t` class would have custom written `__and__` and `__or__` methods which evaluated to the list form for the topic.

Reply:
I just looked up the spec and I don't think my fancy new type idea will work.

Reply:
So far, when people get stuck on events, it's not because they are confused by the topic filters. We should fix the things we know are broken, and keep transparency to the json-rpc layer where we're not sure.

Some examples of things that I have seen trip people up: 1) default fromBlock value is `latest` and the event they're looking for already happened, 2) and disappearing filters on the client.

Some crazy options for the first:
- deviate from the `fromBlock` default in json-rpc, so that it is 0
- *require* that people specify a `fromBlock` when creating a filter?
- always return a result with log queries, with a value like `{fromBlock: 1234, toBlock:1237, events: [], allEvents: [{event-that-did-not-match}]}`, so that it's clear what blocks are being inspected. There are negative performance implications, but maybe you could opt out of that behavior for production.
-

Reply:
I take it back: https://github.com/ethereum/web3.py/issues/188 is an example of confusion on filter topics. Although I'd still argue it's less common that the others.

Reply:
I like the idea of requiring a fromBlock, and including some metadata about the search in the results.  Im less keen on adding the non-matching events to the filter results.  

I think forcing users to think about a choice for the `fromBlock` parameter should be enough. I would prefer keeping the default as `latest` than changing to 0. Typically you need to set up an event listener before catching events. It seems like a more unique situation to listen for events from the past.

Reply:
I added an explanation of how the topic matching works to the doc, so maybe thats enough for #188